contracts/chain-adapters/l2/Ovm_WithdrawalAdapter.sol

```diff
@@ -0,0 +1,138 @@
+    function remoteL1Tokens(address token) external view returns (address);
+
+    // @dev Returns the address for the representation of ETH on the l2.
+    function l2Eth() external view returns (address);
+
+    // @dev Returns the address of the wrapped native token for the L2.
+    function wrappedNativeToken() external view returns (WETH9Interface);
+
+    // @dev Returns the amount of gas the contract allocates for a token withdrawal.
+    function l1Gas() external view returns (uint32);
+}
+
+/**
+ * @title Adapter for interacting with bridges from an OpStack L2 to Ethereum mainnet.
+ * @notice This contract is used to share L2-L1 bridging logic with other L2 Across contracts.
+ */
+contract Ovm_WithdrawalAdapter is WithdrawalAdapter {
+    using SafeERC20 for IERC20;
+
+    // Address for the wrapped native token on this chain. For Ovm standard bridges, we need to unwrap
+    // this token before initiating the withdrawal. Normally, it is 0x42..006, but there are instances
+    // where this address is different.
+    WETH9Interface public immutable wrappedNativeToken;
+    // Address which represents the native token on L2. For OpStack chains, this is generally 0xDeadDeAdde...aDDeAD0000.
+    address public immutable l2Eth;
+    // Stores required gas to send tokens back to L1.
+    uint32 public immutable l1Gas;
+    // Address of the corresponding spoke pool on L2. This is to piggyback off of the spoke pool's supported
+    // token routes/defined token bridges.
+    IOvm_SpokePool public immutable spokePool;
+
+    /*
+     * @notice constructs the withdrawal adapter.
+     * @param _l2Usdc address of native USDC on the L2.
+     * @param _cctpTokenMessenger address of the CCTP token messenger contract on L2.
+     * @param _l2Gateway address of the Optimism ERC20 l2 standard bridge contract.
+     */
+    constructor(
+        IERC20 _l2Usdc,
+        ITokenMessenger _cctpTokenMessenger,
+        address _l2Gateway,
+        IOvm_SpokePool _spokePool
+    ) WithdrawalAdapter(_l2Usdc, _cctpTokenMessenger, _l2Gateway) {
+        spokePool = _spokePool;
+        wrappedNativeToken = spokePool.wrappedNativeToken();
+        l2Eth = spokePool.l2Eth();
+        l1Gas = spokePool.l1Gas();
+    }
+
+    /*
+     * @notice Calls CCTP or the Optimism token gateway to withdraw tokens back to the recipient.
+     * @param recipient L1 address of the recipient.
+     * @param l2TokenAddress address of the l2Token to send back.
+     * @param amountToReturn amount of l2Token to send back.
+     */
+    function withdrawToken(
+        address recipient,
+        address,
+        address l2TokenAddress,
+        uint256 amountToReturn
+    ) public override {
+        // If the token being bridged is WETH then we need to first unwrap it to ETH and then send ETH over the
+        // canonical bridge. On Optimism, this is address 0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000.
```

Comment:
I think we can drop this comment (potentially also from the upstream `Ovm_SpokePool-sol contract`) `0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000` is legacy for Optimism but was deprecated as part of the Bedrock upgrade: https://docs.optimism.io/stack/smart-contracts#deprecated---legacyerc20eth